### PR TITLE
Raise if ChainerX forgets Chainer's comp graph

### PR DESCRIPTION
**Pull request by toslunar**
A tentative fix for #7454.

### CODE REVIEW

**Comment by niboshi** [comment]
TODO: take no-backprop mode into account.

Also please write some tests.

**Comment by toslunar** [comment]
> TODO: take no-backprop mode into account.

Can I assume no-backprop modes are consistent between Chainer and ChainerX?

**Comment by niboshi** [comment]
@toslunar 
Yes, it's user's responsibility to use `chainer.{force,no}_backprop_mode` (which also applies `chainerx.{force,no}_backprop_mode`) when using ChainerX from Chainer.

File: chainer/function_node.py [hunk 1]
```diff
@@ -1204,13 +1205,31 @@ def _extract_apply_in_data(inputs):
...
+                    typ = type(chainer.Variable()).__name__
+                    if isinstance(x, variable.Variable):
+                        typ += '(requires_grad={})'.format(x.requires_grad)
+                        device = x.device
+                    else:
+                        device = backend.get_device_from_array(x)
+                    msg.append(
+                        'input[{}]: {} on {}'.format(
+                            i, typ, device))
+                raise NotImplementedError('\n'.join(msg))
```
Comment (niboshi) [inline chainer/function_node.py]: How about `RuntimeError`? I think it's highly unlikely it will ever be supported.

Comment (toslunar) [reply chainer/function_node.py]: Changed it to `RuntimeError`.

**Comment by niboshi** [comment]
Please add TODO as well.
https://github.com/chainer/chainer/pull/7455#pullrequestreview-250810068

**Comment by toslunar** [comment]
I believe the PR is complete as a bugfix and that it leaves no TODOs.  I'm no longer interested in working on this topic because I wasn't invited to the discussion of #7520.

**Comment by emcastillo** [comment]
#7520 was submitted as a TO-DO on the backlog of chainerx targetting #7515.
I should have checked your PR first, but I didn't see the connection between both issues at first.
I am sorry for creating this missunderstanding.

I believe this PR is more complete and should be merged instead of #7520.

**Comment by toslunar** [comment]
To me, #7454 is a long-standing (> 2 months) issue.  When I asked it internally to members of ChainerX team, their answers are something like below.

- "ChainerX will support backprop through mixed backends.  It's in the ChainerX backlog.  You don't have to work on it."
- "ChainerX team prioritizes the speed of the development.  It's a fair decision to leave the unsafe behavior for a while."

Therefore it was so natural for me to guess that the ChainerX team finally realized the importance of the topic after I publicly wrote it on GitHub.  What I observed recently are

- #7539 was created without a link to/from the PR or #7454.  (*After I found it*, @niboshi and I concluded as https://github.com/chainer/chainer/pull/7539#discussion_r294629181.)
- #7515 was created without a link to/from the PR or #7454.

In general, the lack of acknowledgement makes me angry.  While my view may be affected by my academic career, I'd like to say clear references are important for the work efficiency.

I don't blame @emcastillo because he joined the team this month.  I'm rather happy to see a counter proposal *with a code*.  What I'm not happy with are

- it's likely that someone who knows #7454 assigned #7515 to @emcastillo without mentioning it; and/or
- it's possible that (not only the ChainerX team's decisions are often undocumented, but) they don't remember their own decision.

**Pull request closed without merging**